### ISSUE TITLE
Add clarification to documentation on file-based caching

### ISSUE THREAD

**Issue by FranklinChen** [stefan-k/persistentcache-rs#1]
I couldn't tell from the documentation whether file-based caching results in always loading from disk for every function call. Is this so or is there a secondary cache in memory in case a request occurs in the same running process? The use case I have in mind for a persistent cache is that a function needs to download a resource over the network at a specific URL, and

- the content should be saved on disk, so that if the user runs the program again, the content can be loaded locally from disk keyed off the same URL
- the content should also be cached in memory so that in the same running process, a load of the same URL does not need to go to the persistent cache

**Comment by stefan-k** [comment]
Hi Franklin,
Thank you for your issue! The file-based caching only stores the data on the disk and retrieves it from there, which may be pretty inefficient for your usecase. I've implemented the solution you are looking for in a separate storage with the creative name `FileMemoryStorage`. Alongside caching on disk, it will store the data also in a `HashMap` in memory and retrieve it from there. It will only retrieve from disk if there is no correspoding data in the HashMap. Also, if a value is retrieved from disk, it is stored in the HashMap as well in order to have faster subsequent retrievals. I'm not sure if a HashMap is what you were looking for. For a lot of keys it may not be the most efficient approach. Let me know if you need something different such as LRU.
I published the new version (0.1.6) on crates.io.
I hope this helps, let me know if you run into any problems or if you have further suggestions.

**Comment by FranklinChen** [comment]
Thanks, I'll try this out! I'm expecting not a whole lot of stuff to cache.

**Comment by FranklinChen** [comment]
I finally got around to trying this, but the latest version of stable Rust complains:

```
15 | #![feature(proc_macro)]
   | ^^^^^^^^^^^^^^^^^^^^^^^ help: remove the attribute
   |
   = help: the feature `proc_macro` has been stable since 1.29.0 and no longer requires an attribute to enable
```

**Comment by stefan-k** [comment]
Hi Franklin,

to be honest, in the meantime I came to think that this crate probably isn't that useful (in general) and I haven't been maintaining it for years. I apologize because I should really make this clear in the README and the documentation of this crate. 

I think for your use case it is easier to implement this caching yourself using `serde` to serialize and deserialize the content. The hashed URL could be used as file name, and when the program is started, it checks whether an appropriate file exists and if it does, it loads it from disk and deserializes the content. 

If you do want to use this crate, then I think the problem you mentioned is fixed in the main branch with commit df10a1699c7c35c0298d7fd9f5a8f7e205f3f9b6 . To use this you will have to use the Github URL in your Cargo.toml instead of the version on crates.io. 

You're of course also welcome to fork this project and modify it any way you want (and of course you can also publish it under a new name), but bare in mind that I wrote this in the very early days of my Rust journey and the code is probably just... aweful. ;)

Sorry for the inconvenience.

**Comment by FranklinChen** [comment]
No problem at all! When I returned to the use case I needed, I ended up using the `cached` crate.
